Windows: Add filtering by offset to psscan

adiego8 opened the pull request:
Add capability to windows.psscan to filter by offset. The filter would be used by other plugins to find specific offset using the psscan capabilities, like find hidden processes as a result of dkom for example. The `--offset`  flag argument should represent a physical address space, as supported in volatility2.

The `--offset` flag is included in the following plugins:
  - dlllist
  - handles

adiegovol committed:
Windows: Add filtering by offset to psscan

Add capability to the psscan plugin to filter by specific offset.
The filter would be used by other plugins to find specific offset using the psscan capabilities, like find
hidden processes as a result of some dkom for example.
The `--offset`  flag argument should represent a physical address space.

The flag is included in the following plugins:
  - dlllist
  - handles

ikelos commented:
Thanks very much for this submission, but it looks like there's a few things about the approach that will need changing.  This currently splits the memory layer to look-up an offset for, from the object that the offset came from.  Each object carries around the layer to which its offset relates to.  For example, the same object (the same data) could be at offset `0x1234` in a virtual layer, but `0x5678` in the physical layer, same object, same data, different layers, and different offsets.  If you then look up offset `0x5678` in the virtual layer, you may find the address doesn't exist.

The way to resolve this is to ask the object you're checking for which layer it's in (you'll need a context to do that, the object you're checking should have one, but otherwise you can ask for that at the filer function call stage), determine if that's a physical layer or a virtual one (this is the hardest part, at the moment we just `isinstance` on `IntelLayer` but given we're trying to add AArch64 support, we'll eventually have to change that, for now stick with it as the best way to determine the physical/virtual split), and then return the result (mapping it, if the layer is in fact virtual).  So instead of a simple lambda, this'll need to return a full function which takes an object, finds the layer, determines if the layer is virtual or not and returns the right offset.  No extra variables (like `memory` required.

Also, `create_offset_filter` should either be parameterized to determine whether it should be filtering on physical or virtual addresses, or be renamed `create_physical_offset_filter`, otherwise it might get confusing in the future.

adiegovol committed:
Added support for virtual address filtering and validation of ph/v space

adiego8 commented:
Appreciate the detailed feedback! I added a new [commit](https://github.com/volatilityfoundation/volatility3/pull/1100/commits/5e38ffa811bcf385c2b470351ed2518118c7d875) that I hope addresses all the points you mentioned above based on my understanding, in case I missed something, do not hesitate to let me know.

ikelos reviewed (approved):
Yep, looks much better, thanks.  5:)

ikelos commented on volatility3/framework/plugins/windows/psscan.py:
```diff
@@ -59,6 +59,75 @@
+                    )
+            else:
+                if exclude:
+                    lambda proc: proc.vol.offset == offset
```
So here, you've just made a lambda statement.  I guess you should be assigning it to a variable, so `filter_func =`?

adiego8 replied:
Good catch, thanks! Commit: https://github.com/volatilityfoundation/volatility3/pull/1100/commits/42c6a9ede6bbe3ff52b868144516679753c9fc8c

ikelos commented on volatility3/framework/plugins/windows/psscan.py:
```diff
@@ -59,6 +59,75 @@
+                if exclude:
+                    lambda proc: proc.vol.offset == offset
+                else:
+                    lambda proc: proc.vol.offset != offset
```
Same again.

adiego8 replied:
Commit: https://github.com/volatilityfoundation/volatility3/pull/1100/commits/42c6a9ede6bbe3ff52b868144516679753c9fc8c

ikelos commented on volatility3/framework/plugins/windows/psscan.py:
```diff
@@ -4,7 +4,7 @@
 import datetime
 import logging
-from typing import Iterable, Callable, Optional, Tuple
+from typing import Iterable, Callable, List, Optional, Tuple
```
Looks like this import isn't necessary after all, if you get rid of it that'll resolve this and we can get it all merged.  5:)

adiego8 replied:
Commit: https://github.com/volatilityfoundation/volatility3/pull/1100/commits/42c6a9ede6bbe3ff52b868144516679753c9fc8c

adiego8 replied:
Would you like me to squash additional commits?

eve-mem replied:
I know in the past I've not squashed the extra commits - I think that @ikelos is okay with that. Mostly just reviewing the final product. Although I look forward to maybe being corrected by the man himself! :grin:

ikelos replied:
Yep, no need to squash anything.  Tools are good enough to do the squashing on demand if necessary, and I prefer a full history to see if/how things were changed.  5:)

adiegovol committed:
Remove unnecessary import and correctly filter_func var

ikelos reviewed (approved):
Looks good, thanks!
